packages/flutter/lib/src/rendering/mouse_tracking.dart

```diff
@@ -397,71 +670,146 @@ class MouseTracker extends ChangeNotifier {
-      final Offset lastHoverPosition = previousEvent is PointerHoverEvent ? previousEvent.position : null;
-      final bool pointerHasMoved = lastHoverPosition == null || lastHoverPosition != unhandledEvent.position;
+    if (triggeringEvent is PointerHoverEvent) {
+      final Offset hoverPositionBeforeUpdate = previousEvent is PointerHoverEvent ? previousEvent.position : null;
+      final bool pointerHasMoved = hoverPositionBeforeUpdate == null || hoverPositionBeforeUpdate != triggeringEvent.position;
       // If the hover event follows a non-hover event, or has moved since the
       // last hover, then trigger the hover callback on all annotations.
       // Otherwise, trigger the hover callback only on annotations that it
       // newly enters.
       final Iterable<MouseTrackerAnnotation> hoveringAnnotations = pointerHasMoved ? nextAnnotations.toList().reversed : enteringAnnotations;
       for (final MouseTrackerAnnotation annotation in hoveringAnnotations) {
         if (annotation.onHover != null) {
-          annotation.onHover(unhandledEvent);
+          annotation.onHover(triggeringEvent);
         }
       }
     }
   }
 
-  bool _hasScheduledPostFrameCheck = false;
-  /// Mark all devices as dirty, and schedule a callback that is executed in the
-  /// upcoming post-frame phase to check their updates.
-  ///
-  /// Checking a device means to collect the annotations that the pointer
-  /// hovers, and triggers necessary callbacks accordingly.
-  ///
-  /// Although the actual callback belongs to the scheduler's post-frame phase,
-  /// this method must be called in persistent callback phase to ensure that
-  /// the callback is scheduled after every frame, since every frame can change
-  /// the position of annotations. Typically the method is called by
-  /// [RendererBinding]'s drawing method.
-  void schedulePostFrameCheck() {
-    assert(_duringBuildPhase);
-    assert(!_duringDeviceUpdate);
-    if (!mouseIsConnected)
+  /// MOUSE CURSOR
+
+  /// Returns the current mouse cursor of a device.
+  /// 
+  /// Only valid when asserts are enabled. In release builds, always returns
+  /// null.
+  PreparedMouseCursor debugDeviceCursor(int device) {
+    PreparedMouseCursor result;
+    assert(() {
+      result = _mouseCursorStates[device]?.currentAnnotation?.cursor;
+      return true;
+    }());
+    return result;
+  }
+  final Map<int, _MouseCursorState> _mouseCursorStates = <int, _MouseCursorState>{};
+
+  // Find the mouse cursor, which fallbacks to _fallbackAnnotation.
+  // The `annotations` is the current annotations that the device is
+  // hovering in visual order from front the back.
+  // The return value is never null.
+  MouseTrackerAnnotation _findCursorAnnotation(LinkedHashSet<MouseTrackerAnnotation> annotations) {
+    MouseTrackerAnnotation result;
+    for (final MouseTrackerAnnotation annotation in annotations) {
+      if (annotation.cursor != null) {
+        result = annotation;
```

Comment:
should this break out of the loop if one annotation was found?

Reply:
Can't believe I don't have a test for it. Fixed and added test.